airflow/providers/amazon/aws/operators/sagemaker.py

```diff
@@ -188,7 +193,7 @@ def execute(self, context: 'Context') -> dict:
         )
         if response['ResponseMetadata']['HTTPStatusCode'] != 200:
             raise AirflowException(f'Sagemaker Processing Job creation failed: {response}')
-        return {'Processing': self.hook.describe_processing_job(self.config['ProcessingJobName'])}
+        return {'Processing': serialize(self.hook.describe_processing_job(self.config['ProcessingJobName']))}
```

Comment:
I’m not faimiliar with SageMaker; definitely go for in-place fixes if you feel it is reasonably doable. We can do that _and_ enhancing XCom serialisation; these two don’t exclude each other.

Reply:
Sorry for the delay.  I got caught up helping plan to AWS System Tests migration, and then Airflow Summit.   If I serialize the whole return value, that's changing the return type and I don't know if that's a good idea.  How do you feel about the way it is now?  I think it should be fixed up in the XCOM code as well, I know I ran into similar issue ion the past where a timestamp in an API return breaks an operator's XCOM serialization, so we should look into a larger fix.

Can/should we apply this fix for Sagemaker for the time being and look at XCOM later?

Reply:
What you are doing right now is, strictly speaking, changing the return type, so I don’t see too much of a difference.

Could you share an example output of `self.hook.describe_processing_job(self.config['ProcessingJobName'])`?

Reply:
Yep :)

Reply:
Thanks for the bump :stuck_out_tongue:   For some reason this got stuck as a comment on a pending review instead of a reply.  Strange.  Sorry.

```
{
	'ProcessingInputs': [
		{
			'InputName': 'input', 
			'AppManaged': False, 
			'S3Input': {
				'S3Uri': 's3://bucket/project/preprocessing/input.csv', 
				'LocalPath': '/opt/ml/processing/input', 
				'S3DataType': 'S3Prefix', 
				'S3InputMode': 'File', 
				'S3DataDistributionType': 'FullyReplicated', 
				'S3CompressionType': 'None'
			}
		}
	], 
	'ProcessingOutputConfig': {
		'Outputs': [
			{
				'OutputName': 'output', 
				'S3Output': {
					'S3Uri': 's3://bucket/project/processed-input-data', 
					'LocalPath': '/opt/ml/processing/output', 
					'S3UploadMode': 'EndOfJob'
				},
				'AppManaged': False
			}
		]
	}, 
	'ProcessingJobName': 'project-processing', 
	'ProcessingResources': {
		'ClusterConfig': {
			'InstanceCount': 1, 
			'InstanceType': 'ml.m5.large', 
			'VolumeSizeInGB': 1
		}
	}, 
	'StoppingCondition': {
		'MaxRuntimeInSeconds': 300
	}, 
	'AppSpecification': {
		'ImageUri': '123456789012.dkr.ecr.re-gio-n.amazonaws.com/process_data'}, 
		'RoleArn': 'arn:aws:iam::123456789012:role/service-role/SageMaker-Role', 
		'ProcessingJobArn': 'arn:aws:sagemaker:re-gio-n:123456789012:processing-job/project-processing', 
		'ProcessingJobStatus': 'Completed', 
		'ProcessingEndTime': datetime.datetime(2022, 6, 7, 14, 1, 25, 200000, tzinfo=tzlocal()), 
		'ProcessingStartTime': datetime.datetime(2022, 6, 7, 14, 1, 6, 34000, tzinfo=tzlocal()), 
		'LastModifiedTime': datetime.datetime(2022, 6, 7, 14, 1, 25, 443000, tzinfo=tzlocal()), 
		'CreationTime': datetime.datetime(2022, 6, 7, 13, 57, 29, 487000, tzinfo=tzlocal()), 
		'ResponseMetadata': {
			'RequestId': 'fy2r5b9e-552a-3759-7b2a-d15ef4f32dj3', 
			'HTTPStatusCode': 200, 
			'HTTPHeaders': {
				'x-amzn-requestid': 'fy2r5b9e-552a-3759-7b2a-d15ef4f32dj3', 
				'content-type': 'application/x-amz-json-1.1', 
				'content-length': '1244', 
				'date': 'Tue, 07 Jun 2022 14:01:30 GMT'
			}, 
		'RetryAttempts': 0
	}
}
```

Reply:
@uranusjr Dennis and I discussed this and we feel that in-place replace for Datetime might be more complex hack as compared to the current proposed solution which serializes the entire output while keeping the dictionary type. There multiple Datetimes returned - `describe_training_job` has 9, `describe_processing_job` has 4 and so no  - requiring recursion for an in-place replace. 

Now the choice is between the currently proposed approach or using `AirflowJsonEncoder` in SageMaker operator only. I am still trying to understand how that would be a better approach than this. We'd appreciate your feedback on this as well.

Reply:
@uranusjr Do you have any guidance on what you'd like to see here?  @potiuk didn't seem keen on changing all of XCOM for this and since this is obviously not a universal problem, I kind of agree with that, so what do we need to do to get this unblocked?  What answer would you like to see?

Reply:
I wonder how viable it is if this function recursively look into the dict and convert all datetime values into str (essentially `json.loads(json.dumps(AirflowJsonSerializer))`). Is the dict here generally pretty large or managable?

Reply:
@shubham22 has the numbers in his last comment, 

> There multiple Datetimes returned - describe_training_job has 9, describe_processing_job has 4 and so no - requiring recursion for an in-place replace.

Reply:
I don’t quite understand, why does having multiple datetimes require recursion? JSON-dumping (with an appropriate serializer) converts all datetimes to strs, and loading the JSON back to a dict retains those strs and effectively converts all datetimes.

Reply:
I'll give that a try and see if it works.